graphs/gen/small_world.go

```diff
@@ -0,0 +1,204 @@
+func NavigableSmallWorld(dst GraphBuilder, dims []int, p, q int, r float64, src *rand.Rand) (err error) {
+	if p < 1 {
+		return fmt.Errorf("gen: bad local distance: p=%v", p)
+	}
+	if q < 0 {
+		return fmt.Errorf("gen: bad distant link count: q=%v", q)
+	}
+	if r < 0 {
+		return fmt.Errorf("gen: bad decay constant: r=%v", r)
+	}
+
+	n := 1
+	for _, d := range dims {
+		n *= d
+	}
+	for i := 0; i < n; i++ {
+		if !dst.Has(simple.Node(i)) {
+			dst.AddNode(simple.Node(i))
+		}
+	}
+
+	hasEdge := dst.HasEdge
+	d, isDirected := dst.(graph.Directed)
+	if isDirected {
+		hasEdge = d.HasEdgeFromTo
+	}
+
+	locality := make([]int, len(dims))
+	for i := range locality {
+		locality[i] = p*2 + 1
+	}
+	iterateOver(dims, func(u []int) {
+		uid := idFrom(u, dims)
+		iterateOver(locality, func(delta []int) {
+			d := manhattanDelta(u, delta, dims, -p)
+			if d == 0 || d > p {
+				return
+			}
+			vid := idFromDelta(u, delta, dims, -p)
+			e := simple.Edge{F: simple.Node(uid), T: simple.Node(vid), W: 1}
+			if uid > vid {
+				e.F, e.T = e.T, e.F
+			}
+			if !hasEdge(e.From(), e.To()) {
+				dst.SetEdge(e)
+			}
+			if !isDirected {
+				return
+			}
+			e.F, e.T = e.T, e.F
+			if !hasEdge(e.From(), e.To()) {
+				dst.SetEdge(e)
+			}
+		})
+	})
+
+	defer func() {
+		r := recover()
+		if r != nil {
+			if r != "depleted distribution" {
+				panic(r)
+			}
+			err = errors.New("depleted distribution")
+		}
+	}()
+	w := make([]float64, n)
+	ws := sample.NewWeighted(w, src)
+	iterateOver(dims, func(u []int) {
+		uid := idFrom(u, dims)
+		iterateOver(dims, func(v []int) {
+			d := manhattanBetween(u, v)
+			if d <= p {
+				return
+			}
+			w[idFrom(v, dims)] = math.Pow(float64(d), -r)
+		})
+		ws.ReweightAll(w)
+		for i := 0; i < q; i++ {
+			vid, ok := ws.Take()
+			if !ok {
+				panic("depleted distribution")
+			}
+			e := simple.Edge{F: simple.Node(uid), T: simple.Node(vid), W: 1}
+			if !isDirected && uid > vid {
+				e.F, e.T = e.T, e.F
+			}
+			if !hasEdge(e.From(), e.To()) {
+				dst.SetEdge(e)
+			}
+		}
+		for i := range w {
+			w[i] = 0
+		}
+	})
+
+	return nil
+}
+
+// iterateOver performs an iteration over all dimensions of dims, calling fn
+// for each state. The elements of state must not be mutated by fn.
+func iterateOver(dims []int, fn func(state []int)) {
+	iterator(0, dims, make([]int, len(dims)), fn)
+}
+
+func iterator(d int, dims, state []int, fn func(state []int)) {
+	if d >= len(dims) {
+		fn(state)
+		return
+	}
+	for i := 0; i < dims[d]; i++ {
+		state[d] = i
+		iterator(d+1, dims, state, fn)
+	}
+}
+
+// manhattanBetween returns the Manhattan distance between a and b.
+func manhattanBetween(a, b []int) int {
+	if len(a) != len(b) {
+		panic("gen: unexpected dimension")
+	}
+	var d int
+	for i, v := range a {
+		d += abs(v - b[i])
+	}
+	return d
+}
+
+// manhattanDelta returns the Manhattan distance due to delta over the given
+// dimensions. If delta plus the translation results in out out range of the given
```

Comment:
Yes, though the sentence is horrible. If you can come up with something better, I'll happily substitute.